### PR TITLE
fix(client): no submit on server fail

### LINKED ISSUE #37615: Show notifications for failed backend project submission
Currently when a backend project submission fails, the users are only notified by a brief warning message saying that the server is not being updated and the user is navigated to the next challenge. such as the following:

image

### to reproduce:
to make the backend projects submission fail, please add

```
  componentDidUpdate() {
    this.props.updateProjectForm({});
  }
  componentWillUnmount() {
    this.props.updateProjectForm({});
  }
```
to here https://github.com/freeCodeCamp/freeCodeCamp/blob/740c839f3fcedde5398d9ccc3bf97e93f8ce3502/client/src/templates/Challenges/projects/ProjectForm.js#L29

additionally this #37494 handles the api side of the message on submission failure, so a JSON should be send to the client upon failure.

### expected behaviour:
the user should see the message that was sent from the api-server and not be navigated to the next challenge.

### PR DESCRIPTION
**Pull request by ShaunSHamilton**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #37615 

**22/07/2021** This includes the change in `solution-form` to fail a submission - for testing.

This does not make use of the `updateFetchState` state, as I am leaving this open for some discussion as how best to use it. ?

This was largely possible thanks to the help of these developer: Oliver

### CODE REVIEW

File: client/src/templates/Challenges/redux/completion-epic.js [hunk 1]
```diff
@@ -97,7 +105,13 @@ function submitProject(type, state) {
...
   };
   return postChallenge(update, username).pipe(
-    concat(of(updateSolutionFormValues({})))
+    switchMap(project => {
+      const { type, payload } = project;
+      if (type === challengeActionTypes.updateFailed) {
+        return of(updateFailed(payload));
+      }
+      return of(updateSolutionFormValues({}));
+    })
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/redux/completion-epic.js:114]: I'm not sure about this.  The issue was, I thought, that action with `type == submitChallenge` was being dispatched twice, one of which would trigger the error flash and one of which would close the modal and navigate.

This solves the problem, but only by conditionally dropping the actions that're piped in.  Instead, should we simply only dispatch a single action?

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: client/src/redux/failed-updates-epic.js [hunk 2]
```diff
@@ -41,7 +42,15 @@ function failedUpdateEpic(action$, state$) {
         store.set(key, [...failures, payload]);
       }
     }),
-    map(() => onlineStatusChange(false))
+    map(({ payload }) => {
+      if (payload?.type === 'error') {
```
[the remark this answers is not in the document]
Comment (ShaunSHamilton) [reply client/src/redux/failed-updates-epic.js:46]: Yes. If the user is offline, `payload` does not have a `type` property.

File: client/src/redux/index.js [hunk 3]
```diff
@@ -578,6 +578,17 @@ export const reducer = handleActions(
         }
       };
     },
+    [types.updateFailed]: (state, { payload }) => {
+      return {
+        ...state,
+        updateFetchState: {
+          pending: false,
+          complete: false,
```
[the remark this answers is not in the document]
Comment (ShaunSHamilton) [reply client/src/redux/index.js:586]: Right. That is what I thought, but looking at the other `...FetchState` for a fail, the `complete` is set to false. So, i did not want to change one without changing the other.

File: client/src/templates/Challenges/redux/completion-epic.js (hunk 1 above)
Comment (ShaunSHamilton) [reply client/src/templates/Challenges/redux/completion-epic.js:114]: I am not sure. I thought it needed to dispatch one of those actions, in order for the chain to continue.

File: client/src/redux/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply client/src/redux/index.js:586]: Good point. I've made a note to investigate this, but I agree that we should be consistent.

File: client/src/templates/Challenges/redux/completion-epic.js (hunk 1 above)
Comment (ojeytonwilliams) [reply client/src/templates/Challenges/redux/completion-epic.js:114]: I'm just not convinced two `submitChallenge` actions should be dispatched.

**Comment by ojeytonwilliams** [comment]
Sorry for thinking about this so late, but do we definitely want to wait for a server response before the user can continue?  The reason I ask is that this is not how we do things for the rest of the challenges - we just let them continue and save the submission in local storage.

Do we maybe want to keep the existing behaviour, but make sure that the warning persists?

**Comment by ShaunSHamilton** [comment]
> Do we maybe want to keep the existing behaviour, but make sure that the warning persists?

The way I see it, whilst we might not want to unnecessarily delay a Camper from completing challenges quickly _sans saving_, project submissions are different, and not meant to be "rushed" through. That is, I suspect most Campers want utter-and-complete confirmation their project submission has gone through, and, if not, then they want to try again, or, realise they have submitted something the server is never going to be happy with.

**Comment by ojeytonwilliams** [comment]
That makes sense.  As long as we only change the flow for project submission, that sounds like a plan.

File: client/src/templates/Challenges/redux/completion-epic.js (hunk 1 above)
Comment (ojeytonwilliams) [reply client/src/templates/Challenges/redux/completion-epic.js:114]: @ahmadabdolsaheb am I missing something here?  What we're trying to achieve is:

User clicks submit -> client talks to server -> server responds with error -> user is warned

or

User clicks submit -> client talks to server -> server responds with success -> modal is closed -> navigate to next page

and what I'm worried about is that the action that's dispatched to trigger this *seems* to get dispatched twice.

Comment (ojeytonwilliams) [reply client/src/templates/Challenges/redux/completion-epic.js:114]: After chatting with @ahmadabdolsaheb we realised that ideally it should both ignore duplicate `submitChallenge`s *and* just dispatch a single action.  Reason being rapid submission is almost certainly a bug, so we should code defensively.  And, well, if a single click creates two submissions that's definitely a bug.

**Pull request closed without merging**